components/feature/prompts/src/test/java/mozilla/components/feature/prompts/PromptFeatureTest.kt

```diff
@@ -844,3 +856,153 @@ class PromptFeatureTest {
+            browserState(10),
+            browserState(11),
+            browserState(28),
+            browserState(32),
+            browserState(49),
+            browserState(60),
+            browserState(89)
+            )
+
+        dismissOnPageMostlyLoaded(dialog, stateFlow)
+
+        verify(dialog, times(0)).dismiss()
+    }
+
+    @Test
+    fun `dialog will be dismissed if progress reaches 90%`() = runBlocking {
+        val stateFlow = flowOf(
+            browserState(ContentState(
+                url = "www.mozilla.org",
+                loading = true
+            )),
+            browserState(0),
+            browserState(10),
+            browserState(11),
+            browserState(28),
+            browserState(32),
+            browserState(49),
+            browserState(60),
+            browserState(89),
+            browserState(90)
+        )
+
+        dismissOnPageMostlyLoaded(dialog, stateFlow)
+
+        verify(dialog, times(1)).dismiss()
+    }
+
+    @Test
+    fun `dialog will be dismissed if progress skips past 90%`() = runBlocking {
+        val stateFlow = flowOf(
+            browserState(ContentState(
+                url = "www.mozilla.org",
+                loading = true
+            )),
+            browserState(0),
+            browserState(10),
+            browserState(ContentState(
+                url = "www.mozilla.org",
+                loading = false,
+                progress = 100
+            ))
+        )
+
+        dismissOnPageMostlyLoaded(dialog, stateFlow)
+
+        verify(dialog, times(1)).dismiss()
+    }
+
+    @Test
+    fun `dialog will be dismissed only once if page loads twice`() = runBlocking {
+        val stateFlow = flowOf(
+            browserState(ContentState(
+                url = "www.mozilla.org",
+                loading = true
+            )),
+            browserState(0),
+            browserState(10),
+            browserState(11),
+            browserState(28),
+            browserState(32),
+            browserState(49),
+            browserState(60),
+            browserState(89),
+            browserState(90),
+            browserState(ContentState(
+                url = "www.mozilla.org",
+                loading = false,
+                progress = 100
+            )),
+            browserState(ContentState(
+                url = "www.wikipedia.org",
+                loading = true
+            )),
+            browserState(0),
+            browserState(10),
+            browserState(11),
+            browserState(28),
+            browserState(32),
+            browserState(49),
+            browserState(60),
+            browserState(89),
+            browserState(90),
+            browserState(ContentState(
+                url = "www.wikipedia.org",
+                loading = false,
+                progress = 100
+            ))
+        )
+
+        dismissOnPageMostlyLoaded(dialog, stateFlow)
+
+        verify(dialog, times(1)).dismiss()
+    }
```

Comment:
TODO add test ensures the fun does nothing if the scope has already been canceled.